yarn-project/pxe/src/pxe_service/error_enriching.ts

```diff
@@ -0,0 +1,81 @@
+import { type SimulationError, isNoirCallStackUnresolved } from '@aztec/circuit-types';
+import { AztecAddress, Fr, FunctionSelector, PUBLIC_DISPATCH_SELECTOR } from '@aztec/circuits.js';
+import { type DebugLogger } from '@aztec/foundation/log';
+import { resolveAssertionMessage, resolveOpcodeLocations } from '@aztec/simulator';
+
+import { type ContractDataOracle, type PxeDatabase } from '../index.js';
+
+/**
+ * Adds contract and function names to a simulation error.
+ * @param err - The error to enrich.
+ */
+export async function enrichSimulationError(err: SimulationError, db: PxeDatabase) {
+  // Maps contract addresses to the set of functions selectors that were in error.
+  // Using strings because map and set don't use .equals()
+  const mentionedFunctions: Map<string, Set<string>> = new Map();
+
+  err.getCallStack().forEach(({ contractAddress, functionSelector }) => {
+    if (!mentionedFunctions.has(contractAddress.toString())) {
+      mentionedFunctions.set(contractAddress.toString(), new Set());
+    }
+    mentionedFunctions.get(contractAddress.toString())!.add(functionSelector.toString());
+  });
+
+  await Promise.all(
+    [...mentionedFunctions.entries()].map(async ([contractAddress, selectors]) => {
+      const parsedContractAddress = AztecAddress.fromString(contractAddress);
+      const contract = await db.getContract(parsedContractAddress);
+      if (contract) {
```

Comment:
Maybe log that we attempted to enrich but do now know the contract? Similarly below if we don't find the selector.

Reply:
Done